storage/rocksdb/ha_rocksdb.cc

```diff
@@ -3925,15 +3976,15 @@ class Rdb_transaction {
 
     m_read_opts[USER_TABLE].snapshot = snapshot;
 
-    // TODO: Use snapshot timestamp from rocksdb Snapshot object itself. This
-    // saves the extra call to fetch current time, and allows TTL compaction
-    // (which uses rocksdb timestamp) to be consistent with TTL read filtering
-    // (which uses this timestamp).
-    //
-    // There is no correctness problem though since m_snapshot_timestamp is
-    // generally set after the snapshot has been created, so compaction is
-    // not dropping anything that should have been visible.
-    rdb->GetEnv()->GetCurrentTime(&m_snapshot_timestamp);
+    const auto earliest_ts = get_earliest_snapshot_ts();
+    if (earliest_ts == 0) {
+      auto new_earliest_ts =
+          mysql_bin_log.get_max_write_hlc_nsec() / 1'000'000'000;
+      if (new_earliest_ts == 0) new_earliest_ts = snapshot->GetUnixTime();
+      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
+    } else {
+      assert(earliest_ts <= snapshot->GetUnixTime());
```

Comment:
Abhinav left some comment here:
1. Can we use the mysql_bin_log.get_max_write_hlc_nsec() only when binlog ttl is enabled?
2. For the assertion, when binlog ttl is enabled, we are comparing hlc with wall clock, that doesn't seem correct.
Can we do something like
```
if (earliest_ts == 0) {
      auto new_earliest_ts =
          rdb_is_binlog_ttl_enabled() && mysql_bin_log.get_max_write_hlc_nsec()
              ? mysql_bin_log.get_max_write_hlc_nsec() / 1'000'000'000
              : snapshot->GetUnixTime();
      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
    } else {
      assert(earliest_ts <= rdb_is_binlog_ttl_enabled() &&
                     mysql_bin_log.get_max_write_hlc_nsec()
                 ? mysql_bin_log.get_max_write_hlc_nsec() / 1'000'000'000
                 : snapshot->GetUnixTime());
    }
```

Reply:
This is the last comment needs to be addressed before commit if Luqun and Herman don't have more comments.

Reply:
I am testing the following:
```c++
  [[nodiscard]] static std::int64_t binlog_hlc_or_snapshot_ts(
      const rocksdb::Snapshot &snapshot) {
    if (!rdb_is_binlog_ttl_enabled()) return snapshot.GetUnixTime();
    const auto binlog_max_write_hlc_ns = mysql_bin_log.get_max_write_hlc_nsec();
    return (binlog_max_write_hlc_ns != 0) ? binlog_max_write_hlc_ns
                                          : snapshot.GetUnixTime();
  }

  void snapshot_created(const rocksdb::Snapshot *const snapshot) {
    assert(snapshot != nullptr);

    m_read_opts[USER_TABLE].snapshot = snapshot;

    const auto earliest_ts = get_earliest_snapshot_ts();
    if (earliest_ts == 0) {
      const auto new_earliest_ts = binlog_hlc_or_snapshot_ts(*snapshot);
      Rdb_transaction_list::reinsert_with_snapshot(this, new_earliest_ts);
    } else {
      assert(earliest_ts <= binlog_hlc_or_snapshot_ts(*snapshot));
    }
    m_is_delayed_snapshot = false;
  }
```

Reply:
Ugh, of course with a division by a billion